Add converter for extinction coefficient from CALIOP level 2 aerosol profile (APro) data

weiwilliam opened the pull request:
## Description

This PR adds the converter to process the extinction coefficient from CALIOP Level 2 aerosol profile (APro) data to IODA format.
New variables were introduced.
- extinctionCoefficient
- cloudAerosolDiscrimination


## Issue(s) addressed

Resolves #1711 

## Dependencies

List the other PRs that this PR is dependent on:
- [ ] ...

## Impact

Expected impact on downstream repositories:

## Checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have run the unit tests before creating the PR

weiwilliam committed:
rebuild calipso level 2 extinction converter

weiwilliam committed:
updates

weiwilliam committed:
ready for further test with UFO

weiwilliam committed:
update calipso_l2 converter

weiwilliam committed:
Merge branch 'develop' into feature/calipso_l2_ext

weiwilliam committed:
updates of level2 extinction converter

weiwilliam committed:
updates for qc flags handle

weiwilliam committed:
calipso converter test

weiwilliam committed:
add CAD score and use larger QC value below 8.3km

weiwilliam committed:
updates on caliop l1 and l2 data

weiwilliam committed:
Merge branch 'develop' into feature/caliop_converters

weiwilliam committed:
add converter for CALIOP level2 extinction coefficient in APro product

weiwilliam committed:
update CALIOP converter and the testoutput file

weiwilliam committed:
fix the output of frequency and dimension

weiwilliam committed:
Merge branch 'develop' into feature/caliop_converters

weiwilliam commented:
The JEDI integration test is failed with following message.
```
icpx: error #10105: gcc: core dumped
icpx: warning #10102: unknown signal(-995144544)
icpx: error #10106: Fatal error in gcc, terminated by unknown
clang++: error: unable to make temporary file: No space left on device
clang++: error: no input files
make[2]: *** [ufo/src/ufo/CMakeFiles/ufo.dir/build.make:7411: ufo/src/ufo/CMakeFiles/ufo.dir/operators/marine/insitutemperature/ObsInsituTemperatureTLAD.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
fatal error: error in backend: IO failure on output stream: No space left on device
PLEASE append the compiler options "-save-temps -v", rebuild the application to get the full command which is failing and submit a bug report to https://software.intel.com/en-us/support/priority-support which includes the failing command, input files for the command and the crash backtrace (if any).
```

weiwilliam committed:
update Channels

weiwilliam committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/caliop_converters' into feature/caliop_converters

weiwilliam committed:
change from 3d to 2d for ufo hofx output

weiwilliam committed:
Merge branch 'develop' into feature/caliop_converters

weiwilliam committed:
updates on caliop converters and testoutputs and remove the old files

weiwilliam committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/caliop_converters' into feature/caliop_converters

weiwilliam committed:
add layer thickness which derived from layer midpoint altitude

CoryMartin-NOAA commented on src/compo/caliop_l2_ext2ioda.py:
```diff
@@ -0,0 +1,296 @@
+import os, sys
+
+from pyhdf.HDF import *
+from pyhdf.VS import *
```
This is generally a bad idea to import everything like this

CoryMartin-NOAA commented on src/compo/caliop_l2_ext2ioda.py:
```diff
@@ -0,0 +1,296 @@
+                if varsKeyList[var_keys.index(key)][4]:
+                    self.varAttrs[iodavar, varGroupName]['units'] = varsKeyList[var_keys.index(key)][4]
+
+    def calipso_time2dt(time):
```
Can you add a comment about what the input is expected to look like? This seems really complicated, is there already a way to handle converting this to a datetime object?

weiwilliam replied:
I found it somewhere online with a few tweaks.
I will update it and change the `dateTime` output to `Profile_UTC_Time` rather than `Profile_Time`.
Latter is TAI, which is  37 seconds ahead of UTC since 2017.

CoryMartin-NOAA commented on src/compo/caliop_l2_ext2ioda.py:
```diff
@@ -0,0 +1,296 @@
+        thickness[-1] = alt[-2] - alt[-1]
+
+        # Adjust thickness near 20.2 km because it should be around 180m above and 60m below 20.2km
+        tmpidx = abs(alt-20200).argmin()
```
should be using numpy here

weiwilliam replied:
I updated the code to use numpy functions.

CoryMartin-NOAA commented on src/compo/caliop_l2_ext2ioda.py:
```diff
@@ -0,0 +1,296 @@
+        oldthick = thickness[tmpidx]
+        if alt[tmpidx] > 20200:
+            thickness[tmpidx] = thickness[tmpidx - 1]
+            thickness[tmpidx + 1] = thickness[tmpidx + 1] + abs(thickness[tmpidx] - oldthick)
```
use np.abs on these lines, it should be faster

CoryMartin-NOAA commented on test/testinput/caliop_l2_APro.hdf:
This needs added as a GIT LFS file

weiwilliam replied:
Resolved.

weiwilliam committed:
Merge branch 'develop' into feature/caliop_converters

weiwilliam commented:
Thank you, @CoryMartin-NOAA. I will update the code to address your comments.

weiwilliam committed:
move modis_aod.hdf and caliop_l2_APro.hdf to LFS

weiwilliam committed:
fix the converter and update the testoutput

weiwilliam committed:
update the date_range to cover the whole sample data

eap committed:
Migrate ioda-converters to jedi-ci action (#1696)

Add jedi-ci action

Parent issue: JCSDA-internal/jedi-ci#16
Admin merge justification: JCSDA-internal/jedi-ci#17

weiwilliam committed:
Merge remote-tracking branch 'origin/develop' into feature/caliop_converters

weiwilliam committed:
store profiles indice into sequenceNumber

weiwilliam committed:
update caliop converter and its testoutput

weiwilliam committed:
Merge remote-tracking branch 'origin/develop' into feature/caliop_converters

weiwilliam committed:
Merge remote-tracking branch 'origin/develop' into feature/caliop_converters

weiwilliam committed:
Merge branch 'develop' into feature/caliop_converters

weiwilliam committed:
Merge branch 'develop' into feature/caliop_converters

weiwilliam committed:
Merge branch 'develop' into feature/caliop_converters

weiwilliam committed:
Merge branch 'develop' into feature/caliop_converters

mer-a-o reviewed (approved):
Changes look good to me. I see a small change in `modis_aod.hdf`. Is it because of the change in `.gitattributes`?

weiwilliam commented:
@mer-a-o ~Yes, it should go to LFS. It seems weird now. I will double-check it.~ Confirmed, this branch stores it in LFS.

jeromebarre commented:
Looks good, do you have an associated PR that tests this in UFO?

weiwilliam commented:
@jeromebarre Yes, https://github.com/JCSDA-internal/ufo/pull/3825. But it is in draft PR, because I am looking into whether there is a better approach to load GeoVals for each profile in Load_Atm and Load_Sfc.

jeromebarre commented:
I would refrain on merging this before there is a more matured version of the UFO operator. I have my reservations with using `atmosphereLayerThicknessZ`. I think a top and bottom obs layer pressure should be defined in the ioda file then used in UFO and calculating a more mathematically correct integration on the vertical. Right now the approximaiton that has been coded in in UFO will be innacurate in the case of multiple model levels within an obs layer...

I will add this comment in the UFO PR too.

jeromebarre reviewed (changes requested):
see my comment above
